### PR TITLE
Changes to improve swagger ui

### PR DESCRIPTION
**Pull request by nikhiljindal**
- Separating out version and api handlers into independent web
  services. Moved the version handler to /version and the api handler to /api.
- Registering index Handler on "/" using mux instead of restful WebService since we do not want to surface that in api docs.
- Updating the swagger code to surface swagger files only if config.EnableSwaggerSupport is true.
- Fixed a bug in registering namespace PathParameter, where it was being registered on the WebService instead of being registered on individual Routes.

This is how swagger ui looks now:
y05dp3enbc

oyfda6rnvf

### CODE REVIEW

**Comment by nikhiljindal** [comment]
@bgrant0607

**Comment by bgrant0607** [comment]
Looking at this in detail now. 

I just merged the addition of v1beta3. Could you please rebase and resolve conflicts?

File: pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -264,8 +265,27 @@ func InstallValidator(mux Mux, servers func() map[string]Server) {
 func InstallSupport(container *restful.Container, ws *restful.WebService) {
 	// TODO: convert healthz to restful and remove container arg
 	healthz.InstallHandler(container.ServeMux)
-	ws.Route(ws.GET("/").To(handleIndex))
-	ws.Route(ws.GET("/version").To(handleVersion))
+
+	// Set up the welcome web service.
+	ws.Path("/welcome")
```
Comment (bgrant0607) [inline pkg/apiserver/apiserver.go]: I'd just delete /welcome. It's not useful.

File: pkg/apiserver/index.go [hunk 2]
```diff
@@ -26,10 +26,6 @@ import (
 // handleIndex is the root index page for Kubernetes.
 func handleIndex(req *restful.Request, resp *restful.Response) {
 	// TODO: use restful's Request/Response methods
-	if req.Request.URL.Path != "/" && req.Request.URL.Path != "/index.html" {
```
Comment (bgrant0607) [inline pkg/apiserver/index.go]: Please set up a [redirect](http://golang.org/pkg/net/http/#Redirect) from `/` to something reasonable: maybe healthz?

File: pkg/ui/installsupport.go [hunk 3]
```diff
@@ -32,8 +32,8 @@ func InstallSupport(mux MuxInterface) {
 	prefix := "/static/"
 	mux.Handle(prefix, http.StripPrefix(prefix, fileServer))
 
-	// Expose files in third_party/swagger-ui/ on <host>/swagger-ui/
+	// Expose files in third_party/swagger-ui/ on <host>/
 	fileServer = http.FileServer(&assetfs.AssetFS{Asset: Asset, AssetDir: AssetDir, Prefix: "third_party/swagger-ui"})
-	prefix = "/swagger-ui/"
+	prefix = "/"
```
Comment (bgrant0607) [inline pkg/ui/installsupport.go]: We shouldn't expose this on `/`. For one thing, the swagger support is optional. Speaking of which, this should be optional, similar to: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/master/master.go#L416

**Comment by bgrant0607** [comment]
You plan to create a webservice per REST resource in another PR?

File: pkg/apiserver/apiserver.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/apiserver/apiserver.go]: Done.

File: pkg/apiserver/index.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/apiserver/index.go]: There is not healthz on my localhost.
Have kept the "Welcome to Kubernetes" message for now.
Can change is later if required.

File: pkg/ui/installsupport.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/ui/installsupport.go]: I hadnt realized that swagger support was optional :)
Changed this now.
swagger-ui is exposed on /swagger-ui/ now. And it is exposed only if config.EnableSwaggerSupport is true.

**Comment by nikhiljindal** [comment]
Synced and updated the code for v1beta3.
Re: a webservice per REST resource: as discussed it is not required right now. Can do that later.

**Comment by bgrant0607** [comment]
I searched for all uses of wait_for_url, and most wait for healthz (I don't understand your comment that you don't see healthz). None wait on `/`, so I think we can defer that for now. As discussed in #2057, I think we'll want to return a list of supported paths.

File: test/integration/auth_test.go [hunk 4]
```diff
@@ -270,7 +270,7 @@ func getTestRequests() []struct {
 		// TODO: figure out how to create a minion so that it can successfully proxy/redirect.
 
 		// Non-object endpoints
-		{"GET", "/", "", code200},
+		{"GET", "/welcome", "", code200},
```
Comment (bgrant0607) [inline test/integration/auth_test.go]: Remove this.

File: pkg/master/master.go [hunk 5]
```diff
@@ -421,6 +420,25 @@ func (m *Master) init(c *Config) {
 	m.masterServices.Start()
 }
 
+// Adds a service to return the supported api versions.
+func addApiWebService(container *restful.Container, apiPrefix string, enableV1Beta3 bool) {
+	// TODO: InstallREST should register each version automatically
+	versionHandler := apiserver.APIVersionHandler("v1beta1", "v1beta2")
```
Comment (bgrant0607) [inline pkg/master/master.go]: How about you pass the list of supported versions to this function instead?

File: pkg/apiserver/index.go [hunk 6]
```diff
@@ -17,21 +17,11 @@ limitations under the License.
...
-	// TODO: use restful's Request/Response methods
-	if req.Request.URL.Path != "/" && req.Request.URL.Path != "/index.html" {
-		notFound(resp.ResponseWriter, req.Request)
-		return
-	}
-	resp.ResponseWriter.WriteHeader(http.StatusOK)
-	// TODO: serve this out of a file
-	data := "<html><body>Welcome to Kubernetes</body></html>"
-	fmt.Fprint(resp.ResponseWriter, data)
+func HandleIndex(w http.ResponseWriter, r *http.Request) {
```
Comment (bgrant0607) [inline pkg/apiserver/index.go]: What is the purpose of this change?

File: test/integration/auth_test.go (hunk 4 above)
Comment (nikhiljindal) [inline test/integration/auth_test.go]: Done.

File: pkg/apiserver/index.go (hunk 6 above)
Comment (bgrant0607) [inline pkg/apiserver/index.go]: Other than removing the go-restful args?

File: pkg/master/master.go (hunk 5 above)
Comment (nikhiljindal) [inline pkg/master/master.go]: I can do that. But I am not sure that it is better.
Do you think that passing the list is better?

File: pkg/apiserver/index.go (hunk 6 above)
Comment (nikhiljindal) [inline pkg/apiserver/index.go]: Yes. Mainly removing the go-restful args.
I didnt find the code for checking the path useful. None of the handlers do that check.
I can add it back if you think its useful.

Comment (bgrant0607) [inline pkg/apiserver/index.go]: No, that's fine. Maybe this was a default handler at some point.

File: pkg/master/master.go (hunk 5 above)
Comment (bgrant0607) [inline pkg/master/master.go]: I think the list is better. That would facilitate moving this back to apiserver.go, if we decide to do that, as well as keeping the decision about which versions are exported in one place.

Comment (nikhiljindal) [inline pkg/master/master.go]: Done. Also moved the function to apiserver.

File: pkg/apiserver/apiserver.go [hunk 7]
```diff
@@ -277,6 +311,22 @@ func InstallLogsSupport(mux Mux) {
 	mux.Handle("/logs/", http.StripPrefix("/logs/", http.FileServer(http.Dir("/var/log/"))))
 }
 
+// Adds a service to return the supported api versions.
```
Comment (bgrant0607) [inline pkg/apiserver/apiserver.go]: It would be a little nicer to put AddApiWebService and APIVersionHandler (which could be changed to apiVersionHandler) next to each other, and handleVersion and InstallSupport (which could be renamed InstallVersion) next to each other, but that can be left to a future PR.

**Comment by bgrant0607** [comment]
LGTM
